### ISSUE TITLE
Adding custom command / subcommand to a wp-cli core command

### ISSUE THREAD

**Issue by pierre-dargham** [wp-cli/wp-cli#1767]
**Hi there,**

First of all, thank you for the amazing work you did on wp-cli !

**I'm currently developing a custom command for the main function of our plugin [MultiSite Clone Duplicator](https://wordpress.org/plugins/multisite-clone-duplicator/)**

I successfully apply the [command cookbook](https://github.com/wp-cli/wp-cli/wiki/Commands-Cookbook) recommandations that allowed me to have a command 'site_duplicate'

You can see the code of this command here : https://github.com/pierre-dargham/multisite-clone-duplicator/blob/wp-cli/wp-cli/wp_cli_site_duplicate_command.php

But, as the operation it does (clones a site to another in a multisite network installation) is an atomic operation on sites (we try to provide a 'duplicate' function of the same level as 'create', 'delete', etc.. ), we want to add our command as a subcommand of wp-cli core "site" command.

I succeed in adding my command as 'duplicate' subcommand of 'site' core command with the following code : https://github.com/pierre-dargham/multisite-clone-duplicator/blob/wp-cli/wp-cli/wp_cli_site_duplicate_subcommand.php

**My questions are the following :**

**1. Is it a good way to do it ? Would you accept that community packages add subcommands to wp-cli core commands or would you forbid it ?**

**2. Is there a way / a hook to add my sub command in the command list of the 'site' subcommand, in order to see it when we type 'wp-cli help site' ?**

Thank you for the help.
I wish you success in future developments of this great tool.

**Best regards,**

Pierre Dargham

**Comment by danielbachhuber** [comment]
> 1. Is it a good way to do it ? Would you accept that community packages add subcommands to wp-cli core commands or would you forbid it ?

Sure, seems reasonable.

> 2. Is there a way / a hook to add my sub command in the command list of the 'site' subcommand, in order to see it when we type 'wp-cli help site' ?

Hm, that should just happen automatically. If it doesn't, we can get it fixed.

**Comment by szepeviktor** [comment]
This way it is possible to do it:
```patch
--- o.php	2015-04-20 06:38:40.902687350 +0200
+++ wp_cli_site_duplicate_subcommand.php	2015-04-20 06:37:56.171249700 +0200
@@ -1,18 +1,13 @@
 <?php
 
-require_once WP_CLI_ROOT . '/php/commands/site.php';
-
 /**
  * Implements duplication command.
  */
-
-class Site_Duplicate_Subcommand extends Site_Command  {
+class Site_Duplicate_Subcommand extends WP_CLI_Command  {
 
     /**
      * Duplicate a site in a multisite install.
      *
-     * @alias clone
-     *
      * ## OPTIONS
      *
      * --slug=<slug>
@@ -30,27 +25,29 @@
      * --network_id=<network-id>
      * : Network to associate new site with. Defaults to current network (typically 1).
      *
-     * --private
+	 * [--private]
      * : If set, the new site will be non-public (not indexed)
      *
-     * --porcelain
+	 * [--porcelain]
      * : If set, only the site id will be output on success.
      *
-     * --v
+	 * [--v]
      * : If set, print more details about the new site (Verbose mode). Do not work if --procelain is set.
      *
-     * --do_not_copy_files
+	 * [--do_not_copy_files]
      * : If set, files of the duplicated site will not be copied.
      *
-     * --keep_users
+	 * [--keep_users]
      * : If set, the new site will have the same users as the duplicated site.
      *
-     * --log=<dir_path>
+	 * [--log=<dir_path>]
      * : If set, a log will be written in this directory (please check this directory is writable).
      *
+	 * @alias clone
+	 *
      * @synopsis --slug=<slug> --source=<site_id> [--title=<title>] [--email=<email>] [--network_id=<network-id>] [--private] [--porcelain] [--v] [--do_not_copy_files] [--keep_users] [--log=<dir_path>]
      */
-    public function duplicate( $_, $assoc_args ) {
+	public function __invoke( $_, $assoc_args ) {
         if ( !is_multisite() ) {
             WP_CLI::error( 'This is not a multisite install.' );
         }
@@ -190,4 +187,4 @@
 
 }
 
-WP_CLI::add_command( 'site', 'Site_Duplicate_Subcommand' );
\ No newline at end of file
+WP_CLI::add_command( 'site duplicate', 'Site_Duplicate_Subcommand' );
```

**Comment by szepeviktor** [comment]
@pierre-dargham Please watch yout coding style.
Rename `wp_cli_site_duplicate_subcommand.php` to `wp-cli-site-duplicate-subcommand.php`
Check for your class: `is_class( 'MUCD_Functions' )`
Your sanitization may fail e.g. on `sité~slug`.
Please consider adding tests.

**Comment by pierre-dargham** [comment]
Thank you for the answers !

I followed some of @szepeviktor recommandations, it looks better now : https://github.com/pierre-dargham/multisite-clone-duplicator/blob/wp-cli/wp-cli/wp-cli-site-duplicate-subcommand.php

@danielbachhuber : my subcommand works fine, when I type `wp site duplicate <args>`, it runs. But `wp site help` still doesn't mention my subcommand. Do you think I did something wrong, or could it be some bug in wp-cli core ?

Best regards,

Pierre

**Comment by szepeviktor** [comment]
Here you find the reason:
https://github.com/pierre-dargham/multisite-clone-duplicator/blob/wp-cli/wp-cli/wp-cli-site-duplicate-subcommand.php#L7
I think wp-cli does not load WP on `help`.

**Comment by pierre-dargham** [comment]
After looking for the problem in wp-cli php files, I'm pretty sure custom subcommands does not show on `wp help <command>` because it is runned before `wp-settings-cli.php` is loaded.

Can we do something about it ? I opened a [pull-request](https://github.com/wp-cli/wp-cli/pull/1793) that solve the problem, can you see if you agree with the changes I made ?

**Comment by danielbachhuber** [comment]
> Can we do something about it ?

Yep, after 0.19.0 is released #1795

**Comment by pierre-dargham** [comment]
Thank you ! We'll be waiting for 0.19.0 !

Best regards

Pierre
